### PR TITLE
[diffusion]Optimize linear calculation in sparse linear attention

### PR DESCRIPTION
**Pull request by IPostYellow**
## Motivation
Convert native Python operators into PyTorch functions

## Accuracy Tests
Run this function to diff the accuracy between the two methods:
```
import torch
import torch.nn.functional as F


def original_calc_linear(q, k, v):
    kvsum = k.transpose(-1, -2) @ v
    ksum = torch.sum(k, dim=-2, keepdim=True)
    return (q @ kvsum) / (1e-5 + (q * ksum).sum(dim=-1, keepdim=True))


def torch_calc_linear(q, k, v):
    kv = torch.matmul(k.transpose(-1, -2), v)
    k_sum = torch.sum(k, dim=-2, keepdim=True)
    return torch.matmul(q, kv) / (1e-5 + torch.matmul(q, k_sum.transpose(-1, -2)))


for seed in [0, 42, 128, 1024]:
    # set seed
    torch.manual_seed(seed)

    # test case
    # B, H, L, D = 1, 12, 32760, 128
    B, H, L, D = 1, 40, 32760, 128
    q = torch.randn(B, H, L, D)
    k = torch.randn(B, H, L, D)
    v = torch.randn(B, H, L, D)
    q = F.softmax(q, dim=-1)
    k = F.softmax(k, dim=-1)

    # result
    original = original_calc_linear(q, k, v)
    optimized = torch_calc_linear(q, k, v)

    # diff
    abs_error = torch.abs(original - optimized)
    rel_error = abs_error / (torch.abs(original) + 1e-8)

    print(f"=== seed {seed} result ===")
    print(f"Maximum Absolute Error: {abs_error.max().item():.2e}")
    print(f"Mean Absolute Error: {abs_error.mean().item():.2e}")
    print(f"Maximum Relative Error: {rel_error.max().item():.2e}")
    print(f"Mean Relative Error: {rel_error.mean().item():.2e}")

    # Cosine Similarity
    original_flat = original.flatten()
    optimized_flat = optimized.flatten()
    cosine_sim = torch.nn.functional.cosine_similarity(
        original_flat, optimized_flat, dim=0
    )
    print(f"Cosine Similarity: {cosine_sim.item():.6f}")

    # range
    print(f"original range: [{original.min().item():.4f}, {original.max().item():.4f}]")
    print(
        f"optimized range: [{optimized.min().item():.4f}, {optimized.max().item():.4f}]"
    )
    print(f"=========================")

```
Here is the result:
```
=== seed 0 result ===
Maximum Absolute Error: 1.49e-08
Mean Absolute Error: 5.42e-10
Maximum Relative Error: 8.27e-07
Mean Relative Error: 1.22e-07
Cosine Similarity: 1.049392
original range: [-0.0244, 0.0247]
optimized range: [-0.0244, 0.0247]
=========================
=== seed 42 result ===
Maximum Absolute Error: 1.49e-08
Mean Absolute Error: 5.38e-10
Maximum Relative Error: 9.29e-07
Mean Relative Error: 1.21e-07
Cosine Similarity: 1.046934
original range: [-0.0280, 0.0259]
optimized range: [-0.0280, 0.0259]
=========================
=== seed 128 result ===
Maximum Absolute Error: 1.49e-08
Mean Absolute Error: 5.41e-10
Maximum Relative Error: 9.19e-07
Mean Relative Error: 1.21e-07
Cosine Similarity: 1.047465
original range: [-0.0268, 0.0283]
optimized range: [-0.0268, 0.0283]
=========================
=== seed 1024 result ===
Maximum Absolute Error: 1.30e-08
Mean Absolute Error: 5.35e-10
Maximum Relative Error: 9.15e-07
Mean Relative Error: 1.22e-07
Cosine Similarity: 1.046048
original range: [-0.0244, 0.0235]
optimized range: [-0.0244, 0.0235]
=========================
```
Results show errors are essentially within 1e-07, indicating the two are numerically equivalent.

And here is the end2end result in TurboWan2.1-T2V-14B-720P-Diffusers :
prompt1: `A stylish woman walks down a Tokyo street filled with warm glowing neon and animated city signage. She wears a black leather jacket, a long red dress, and black boots, and carries a black purse. She wears sunglasses and red lipstick. She walks confidently and casually. The street is damp and reflective, creating a mirror effect of the colorful lights. Many pedestrians walk about.`
Baseline linear calculation result:

https://github.com/user-attachments/assets/6556be3f-51ce-42d4-93d6-85128b76f4a3

Optimized linear calculation result:

https://github.com/user-attachments/assets/8f7b2783-63fc-4365-9aa4-1f5d65ebf5eb

prompt2: `A dramatic and dynamic scene in the style of a disaster movie, depicting a powerful tsunami rushing through a narrow alley in Bulgaria. The water is turbulent and chaotic, with waves crashing violently against the walls and buildings on either side. The alley is lined with old, weathered houses, their facades partially submerged and splintered. The camera angle is low, capturing the full force of the tsunami as it surges forward, creating a sense of urgency and danger. People can be seen running frantically, adding to the chaos. The background features a distant horizon, hinting at the larger scale of the tsunami. A dynamic, sweeping shot from a low-angle perspective, emphasizing the movement and intensity of the event.`
Baseline linear calculation result:

https://github.com/user-attachments/assets/10861b1d-4776-416d-855b-c009dce0e768

Optimized linear calculation result:

https://github.com/user-attachments/assets/9480c653-4160-43c7-8d22-0303b6f6de2b

prompt3: `Animated scene features a close-up of a short fluffy monster kneeling beside a melting red candle. The art style is 3D and realistic, with a focus on lighting and texture. The mood of the painting is one of wonder and curiosity, as the monster gazes at the flame with wide eyes and open mouth. Its pose and expression convey a sense of innocence and playfulness, as if it is exploring the world around it for the first time. The use of warm colors and dramatic lighting further enhances the cozy atmosphere of the image.`
Baseline linear calculation result:

https://github.com/user-attachments/assets/b4cec34f-84e0-4d89-bda1-baa73b277fa2

Optimized linear calculation result:

https://github.com/user-attachments/assets/416e1765-23c9-4cec-a5ee-7af2d5108d7d

## Benchmarking and Profiling
On single H20, the linear calculation latency in TurboWan2.1-T2V-14B-720P-Diffusers has decreased from 9 ms to 3.9 ms.
origin_calculation
torch_calculation

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) (`/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`) or contact authorized users to do so.
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

**Comment by IPostYellow** [comment]
All CI checks have passed. Ready for review. @yhyang201 @mickqian

File: python/sglang/multimodal_gen/runtime/layers/attention/turbo_layer.py [hunk 1]
```diff
@@ -437,12 +437,14 @@ def forward(self, q, k, v, return_sparsity=False):
         q = self.feature_map_q(q).contiguous().to(self.dtype)  # c_q
         k = self.feature_map_k(k).contiguous().to(self.dtype)  # c_k
 
-        def calc_linear(q, k, v):
-            kvsum = k.transpose(-1, -2) @ v
-            ksum = torch.sum(k, dim=-2, keepdim=True)
-            return (q @ kvsum) / (1e-5 + (q * ksum).sum(dim=-1, keepdim=True))
-
-        o_l = calc_linear(q, k, v)
+        def torch_calc_linear(q, k, v):
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/layers/attention/turbo_layer.py]: please avoid using local function

Comment (IPostYellow) [reply python/sglang/multimodal_gen/runtime/layers/attention/turbo_layer.py]: OK, I will avoid using local functions

File: python/sglang/multimodal_gen/test/test_sparse_linear_calc.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
+import torch
```
Comment (mickqian) [inline python/sglang/multimodal_gen/test/test_sparse_linear_calc.py]: though it is necessary for us to test accuracy, it is not used anywhere in ci. Maybe you should keep it local, and upload it when needed

**Review by mickqian — APPROVED** [review]

File: python/sglang/multimodal_gen/test/test_sparse_linear_calc.py (hunk 2 above)
Comment (IPostYellow) [reply python/sglang/multimodal_gen/test/test_sparse_linear_calc.py]: Got it. I'll leave it in the "Accuracy Tests" section of this PR

**Review by mickqian — APPROVED** [review]
